### PR TITLE
[c++] Refine concurrency configuration (`soma.compute_concurrency_level`)

### PR DESCRIPTION
**Pull request by bkmartinjr**
**Issue and/or context:**

Introduces direct control over the native thread pool concurrency level in SOMA. Previously, it was indirectly controlled by the TileDB Core `sm.compute_currency_control` configuration, and would create a thread pool that was 1/2 the size of that value.

**Changes:**

This PR introduces a new context configuration setting:  `soma.compute_currency_level`, which defaults to the host CPU count. It can be set with an numeric value,eg., in Python:

```python
context = tiledbsoma.SOMATileDBContext(tiledb_config={"soma.compute_currency_level": "20"})
```

**Notes for Reviewer:**

sc-54891

the PR also adds:
* improved error handling for malformed config values
* bound the value between 1 and 1024

### CODE REVIEW

File: apis/python/tests/test_context.py [hunk 1]
```diff
@@ -137,7 +137,7 @@ def test_replace_config_after_construction():
 def test_malformed_concurrency_config_value():
     import numpy as np
 
-    with pytest.raises(tiledbsoma.SOMAError):
+    with pytest.raises((tiledbsoma.SOMAError, RuntimeError)):
```
Comment (johnkerl) [inline apis/python/tests/test_context.py:140]: I'll help you fix this -- already done on https://github.com/single-cell-data/TileDB-SOMA/pull/2963

* https://github.com/single-cell-data/TileDB-SOMA/blob/c08b5cf544040a9601fa8c19496ff4af6c7f2c8a/apis/python/src/tiledbsoma/pytiledbsoma.cc#L36-L45
* https://github.com/single-cell-data/TileDB-SOMA/blob/c08b5cf544040a9601fa8c19496ff4af6c7f2c8a/apis/python/src/tiledbsoma/soma_array.cc#L348-L352 (one example callsite)

Comment (johnkerl) [reply apis/python/tests/test_context.py:140]: Also
https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html

Comment (bkmartinjr) [reply apis/python/tests/test_context.py:142]: Broader issue - filed as sc-54978

**Review by johnkerl — APPROVED** [review]
LGTM but I do defer to @nguyenv

File: apis/python/tests/test_context.py [hunk 2]
```diff
@@ -131,3 +132,16 @@ def test_replace_config_after_construction():
             new_tdb_ctx = new_soma_ctx.tiledb_ctx
         mock_ctx.assert_called_once()
         assert new_tdb_ctx.config()["vfs.s3.region"] == "us-west-2"
+
+
+def test_malformed_concurrency_config_value():
+    import numpy as np
```
Comment (nguyenv) [inline apis/python/tests/test_context.py:138]: Why import in the test instead of up top?

**Review by nguyenv — APPROVED** [review]

File: apis/python/tests/test_context.py (hunk 2 above)
Comment (bkmartinjr) [reply apis/python/tests/test_context.py:138]: Just trying to isolate test dependencies.  But not a major issue either way....

**Pull request merged**